bokehjs/src/coffee/renderer/tile/mercator_tile_source.coffee

```diff
@@ -0,0 +1,241 @@
+
+  tms_to_wmts: (x, y, z) ->
+    '''Note this works both ways'''
+    return [x, 2 ** z - 1 - y, z]
+
+  wmts_to_tms: (x, y, z) ->
+    '''Note this works both ways'''
+    return [x, 2 ** z - 1 - y, z]
+
+  pixels_to_meters: (px, py, level) ->
+    res = @get_resolution(level)
+    mx = px * res - @get('x_origin_offset')
+    my = py * res - @get('y_origin_offset')
+    return [mx, my]
+
+  meters_to_pixels: (mx, my, level) ->
+    res = @get_resolution(level)
+    px = (mx + @get('x_origin_offset')) / res
+    py = (my + @get('y_origin_offset')) / res
+    return [px, py]
+
+  pixels_to_tile: (px, py) ->
+    tx = Math.max(Math.ceil(px / parseFloat(@get('tile_size'))) - 1, 0)
+    ty = Math.max(Math.ceil(py / parseFloat(@get('tile_size'))) - 1, 0)
+    return [tx, ty]
+
+  pixels_to_raster: (px, py, level) ->
+    mapSize = @get('tile_size') << level
+    return [px, mapSize - py]
+
+  meters_to_tile: (mx, my, level) ->
+    [px, py] = @meters_to_pixels(mx, my, level)
+    return @pixels_to_tile(px, py)
+
+  get_tile_meter_bounds: (tx, ty, level) ->
+    # expects tms styles coordinates (bottom-left origin)
+    [xmin, ymin] = @pixels_to_meters(tx * @get('tile_size'), ty * @get('tile_size'), level)
+    [xmax, ymax] = @pixels_to_meters((tx + 1) * @get('tile_size'), (ty + 1) * @get('tile_size'), level)
+
+    if xmin? and ymin? and xmax? and ymax?
+      return [xmin, ymin, xmax, ymax]
+    else
+      return undefined
+
+  get_tile_geographic_bounds: (tx, ty, level) ->
+    bounds = @get_tile_meter_bounds(tx, ty, level)
+    [minLon, minLat, maxLon, maxLat] = @utils.meters_extent_to_geographic(bounds)
+    return [minLon, minLat, maxLon, maxLat]
+
+  get_tiles_by_extent: (extent, level, tile_border=1) ->
+    # unpack extent and convert to tile coordinates
+    [xmin, ymin, xmax, ymax] = extent
+    [txmin, tymin] = @meters_to_tile(xmin, ymin, level)
+    [txmax, tymax] = @meters_to_tile(xmax, ymax, level)
+
+    # add tiles which border
+    txmin -= tile_border
+    tymin -= tile_border
+    txmax += tile_border
+    tymax += tile_border
+
+    tiles = []
+    for ty in [tymax..tymin] by -1
+      for tx in [txmin..txmax] by 1
+        tiles.push([tx, ty, level, @get_tile_meter_bounds(tx, ty, level)])
+        tiles = @sort_tiles_from_center(tiles, [txmin, tymin, txmax, tymax])
+    return tiles
+
+  quadkey_to_tile_xyz: (quadKey) ->
+    '''
+    Computes tile x, y and z values based on quadKey.
+    '''
+    tileX = 0
+    tileY = 0
+    tileZ = quadKey.length
+    for i in [tileZ...0] by -1
+
+      value = quadKey.charAt(tileZ - i)
+      mask = 1 << (i - 1)
+
+      if value == '0'
+        continue
+
+      else if value == '1'
+        tileX |= mask
+
+      else if value == '2'
+        tileY |= mask
+
+      else if value == '3'
+        tileX |= mask
+        tileY |= mask
+
+      else
+        throw new TypeError("Invalid Quadkey: " + quadKey)
+
+    return [tileX, tileY, tileZ]
+
+  tile_xyz_to_quadkey: (x, y, z) ->
```

Comment:
i think the quadkey specific stuff should go in the quadkey tile source - this method is only called from there as far as i can see.